papermill/engines.py

```diff
@@ -0,0 +1,90 @@
+import copy
+import datetime
+
+from .preprocess import PapermillExecutePreprocessor
+from .iorw import write_ipynb
+
+class PapermillEngines(object):
+    def __init__(self):
+        self._engines = {}
+
+    def register(self, name, engine):
+        self._engines[name] = engine
+
+    def get_engine(self, name=None):
+        engine = self._engines.get(name)
+        if not engine:
+            raise PapermillException("No engine named '{}' found".format(name))
+        return engine
+
+    def execute_notebook_with_engine(self, engine_name, nb, kernel_name, **kwargs):
+        return self.get_engine(engine_name).timed_execute_notebook(nb, kernel_name, **kwargs)
+
+
+class EngineBase(object):
+    @classmethod
+    def timed_execute_notebook(cls, nb, kernel_name, **kwargs):
+        """Wraps the execute_notebook function with some metadata timing assignments"""
+        t0 = datetime.datetime.utcnow()
+        nb = cls.execute_notebook(nb, kernel_name, **kwargs)
+        t1 = datetime.datetime.utcnow()
+
+        nb.metadata.papermill['start_time'] = t0.isoformat()
+        nb.metadata.papermill['end_time'] = t1.isoformat()
+        nb.metadata.papermill['duration'] = (t1 - t0).total_seconds()
+
+        return nb
+
+    @classmethod
+    def extract_exception(cls, nb):
+        """Reusable by most engines"""
+        return any([cell.metadata.papermill.get('exception') for cell in nb.cells])
+
+    @classmethod
+    def execute_notebook(cls, kernel_name, **kwargs):
+        raise NotImplementedError("'execute_notebook' is not implemented for this engine")
+
+
+class NBConvertEngine(EngineBase):
+    @classmethod
+    def execute_notebook(
+        cls,
+        nb,
+        kernel_name,
+        output_path=None,
+        progress_bar=True,
```

Comment:
Was wondering if you want to leave the progress bar to each engine or have it as something "external".

I might have a slight preference for making the progress bar something that is provided (or not) by the caller of the engine. The engine gets passed a callback that it can (or not) call every time it makes progress. Something along the lines of https://github.com/tqdm/tqdm#manual

Maybe something for a new PR as it would require surgery on the executor.

Reply:
I think I generally agree with this idea. Let me play with it some here. Worse case we can do a second pass PR for items like this.